hudi-sync/hudi-sync-common/src/main/java/org/apache/hudi/sync/common/util/ManifestFileUtil.java

```diff
@@ -0,0 +1,142 @@
+ * limitations under the License.
+ */
+
+package org.apache.hudi.sync.common.util;
+
+import org.apache.hudi.common.config.HoodieMetadataConfig;
+import org.apache.hudi.common.config.SerializableConfiguration;
+import org.apache.hudi.common.engine.HoodieLocalEngineContext;
+import org.apache.hudi.common.fs.FSUtils;
+import org.apache.hudi.common.model.HoodieBaseFile;
+import org.apache.hudi.common.table.HoodieTableMetaClient;
+import org.apache.hudi.common.util.FileIOUtils;
+import org.apache.hudi.common.util.Option;
+import org.apache.hudi.common.util.ValidationUtils;
+import org.apache.hudi.exception.HoodieException;
+import org.apache.hudi.metadata.HoodieMetadataFileSystemView;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.Path;
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static org.apache.hudi.common.config.HoodieMetadataConfig.DEFAULT_METADATA_ENABLE_FOR_READERS;
+import static org.apache.hudi.common.config.HoodieMetadataConfig.ENABLE;
+
+public class ManifestFileUtil {
+  private static final long serialVersionUID = 1L;
+  private static final Logger LOG = LogManager.getLogger(ManifestFileUtil.class);
+  private static final String MANIFEST_FOLDER_NAME = "manifest";
+  private static final String MANIFEST_FILE_NAME = "latest-snapshot.csv";
+  private static final String DELIMITER = "\n";
+  private SerializableConfiguration hadoopConf;
+  private String basePath;
+  private transient HoodieLocalEngineContext engineContext;
+  private HoodieTableMetaClient metaClient;
+
+  private ManifestFileUtil(Configuration conf, String basePath) {
+    this.hadoopConf = new SerializableConfiguration(conf);
+    this.basePath = basePath;
+    this.engineContext = new HoodieLocalEngineContext(conf);
+    this.metaClient = HoodieTableMetaClient.builder().setConf(hadoopConf.get()).setBasePath(basePath).setLoadActiveTimelineOnLoad(true).build();
+  }
+
+  public ManifestFileUtil() {
+  }
+
+  public synchronized void writeManifestFile() {
+    try {
+      Path manifestFilePath = new Path(getManifestFolder(), MANIFEST_FILE_NAME);
+      Option<byte[]> content = Option.of(fetchLatestBaseFilesForAllPartitions().collect(Collectors.joining(DELIMITER)).getBytes());
+      FileIOUtils.createFileInPath(metaClient.getFs(), manifestFilePath, content);
+    } catch (Exception e) {
+      String msg = "Error writing manifest file";
+      LOG.error(msg, e);
+      throw new HoodieException(msg, e);
+    }
+  }
+
+  public Stream<String> fetchLatestBaseFilesForAllPartitions() {
+    try {
+      HoodieMetadataConfig metadataConfig = buildMetadataConfig(hadoopConf.get());
+      HoodieMetadataFileSystemView fsView = new HoodieMetadataFileSystemView(engineContext, metaClient, metaClient.getActiveTimeline().getCommitsTimeline().filterCompletedInstants(), metadataConfig);
```

Comment:
you should probably create a new HoodieMetadataFileSystemView within the lambda below, also create HoodieLocalEngineContext there instead of keep 1 for the entire object. HoodieLocalEngineContext is not serializeable.